CODINGSTYLE.md

```diff
@@ -0,0 +1,335 @@
+# Coding Style Guide
+
+The Scala.js project has pretty drastic rules in terms of coding style.
+This document tries to document them as much as possible.
+
+A *few* of these rules are checked automatically using Scalastyle, but most of them are too complex to teach to an automated tool.
+
+In some cases, [@sjrd](https://github.com/sjrd) will just say something else on a code review.
+His opinion on a direct piece of code always takes precedence over this document.
+Feel free to point out the conflict, though: the document can be improved.
+
+
+## General style (whitespaces, braces, etc.)
+
+### Tabs, new lines, and eofs
+
+* Files must not contain tabs
+* New lines are UNIX style (\n)
+* A final new line must be present at the end of file
+* There must not be whitespace characters at the end of a line
+
+### Line length
+
+Lines should be limited at 80 characters.
+In some cases, if breaking a line makes it significantly less readable, it can go up to 120 characters.
+
+Rationale: when reviewing on GitHub, only 120 characters are visible; when reviewing on a mobile phone, only 80 characters are visible.
+And we do review on mobile phone quite a lot.
+
+#### Where to break a line
+
+A line can be broken after either a `,` or a `(`, or possibly after a binary operator in a long expression.
+
+### Indentation
+
+In general, indentation is 2 spaces, except continuation lines, which are indented 4 spaces.
+A continuation line is a line appearing because we broke something that should have been on one line for line length reasons.
+Typically, this means inside parentheses (formal or actual parameters of methods and constructors), and a long `extends` clause.
+
+Note that breaking a line right after the `=` sign of an initialization or assignment is *not* considered a continuation line.
+
+### Blank lines
+
+* Never put two blank lines in a row
+* (Almost) always put a blank line between two declarations in a class (exceptions are several related consecutive ScalaDocless private fields)
+* Always put blank lines around a `case` whose body spans several lines
+* Insert blank lines at will between logical blocks of statements in a method
+
+### Braces
+
+Braces should not be used for one-liner methods, nor for two-liner `if/else` methods.
+Similarly, you should not use braces for the body of a `while`/`if`/`for` that is a one-liner (but you should if it is a two-liner `if/else`).
+
+```scala
+def abs(x: Int): Int =
+  if (x >= 0) x
+  else -x
+
+def clamp(v: Int, min: Int, max: Int): Int = {
+  if (v < min) min
+  else if (v > max) max
+  else v
+}
+```
+
+Every expression that spans more than one line should be enclosed in braces.
+For example, this is wrong:
```

Comment:
Right. Somehow a two-liner if/else has the status of a one-liner in most cases.

Not always, though. As a counter-example, this is still invalid:
```scala
if (x > 0)
  if (x < 5) x
  else x * 2
else
  x + 4
```